drivers/photo-electronic/adpd188/adpd188.c

```diff
@@ -0,0 +1,876 @@
+{
+	int32_t ret;
+	uint16_t reg_data;
+
+	if(!flags)
+		return SUCCESS;
+
+	ret = adpd188_reg_read(dev, ADPD188_REG_STATUS, &reg_data);
+	if(ret != SUCCESS)
+		return FAILURE;
+
+	/*
+	 * If an interrupt is not to be cleared, but happens to be asserted write 0
+	 * to that spot to not clear it unintentionally.
+	 */
+	if(flags & ADPD188_SLOTA_INT)
+		reg_data |= ADPD188_STATUS_SLOTA_INT_MASK;
+	else
+		reg_data &= ~ADPD188_STATUS_SLOTA_INT_MASK;
+	if(flags & ADPD188_SLOTB_INT)
+		reg_data |= ADPD188_STATUS_SLOTB_INT_MASK;
+	else
+		reg_data &= ~ADPD188_STATUS_SLOTB_INT_MASK;
+
+	return adpd188_reg_write(dev, ADPD188_REG_STATUS, reg_data);
+}
+
+/**
+ * @brief Enable the slot and FIFO interrupt flags.
+ * @param dev - The ADPD188 descriptor.
+ * @param flags - The ORed value of the flags that need to be enabled. If no
+ *                interrupt needs to be enabled the value will be zero.
+ *                Otherwise do a logical OR between the flags that need to be
+ *                enabled:
+ *                - ADPD188_SLOTA_INT;
+ *                - ADPD188_SLOTB_INT;
+ *                - ADPD188_FIFO_INT.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd188_interrupt_en(struct adpd188_dev *dev, uint8_t flags)
+{
+	int32_t ret;
+	uint16_t reg_data;
+
+	ret = adpd188_reg_read(dev, ADPD188_REG_INT_MASK, &reg_data);
+	if(ret != SUCCESS)
+		return FAILURE;
+	if(flags & ADPD188_SLOTA_INT)
+		reg_data &= ~ADPD188_INT_MASK_SLOTA_INT_MASK_MASK;
+	if(flags & ADPD188_SLOTB_INT)
+		reg_data &= ~ADPD188_INT_MASK_SLOTB_INT_MASK_MASK;
+	if(flags & ADPD188_FIFO_INT)
+		reg_data &= ~ADPD188_INT_MASK_FIFO_INT_MASK_MASK;
+
+	return adpd188_reg_write(dev, ADPD188_REG_INT_MASK, reg_data);
+}
+
+/**
+ * @brief Setup drive and polarity of the GPIOs. Also enable GPIO if necessary.
+ * @param dev - The ADPD188 descriptor.
+ * @param config - Configuration structure of the GPIO.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd188_gpio_setup(struct adpd188_dev *dev,
+			   struct adpd188_gpio_config config)
+{
+	int32_t ret;
+	uint16_t reg_data;
+
+	ret = adpd188_reg_read(dev, ADPD188_REG_GPIO_DRV, &reg_data);
+	if(ret != SUCCESS)
+		return FAILURE;
+	if(config.gpio_id == 0) {
+		if(config.gpio_pol)
+			reg_data |= ADPD188_GPIO_DRV_GPIO0_POL_MASK;
+		else
+			reg_data &= ~ADPD188_GPIO_DRV_GPIO0_POL_MASK;
+		if(config.gpio_drv)
+			reg_data |= ADPD188_GPIO_DRV_GPIO0_DRV_MASK;
+		else
+			reg_data &= ~ADPD188_GPIO_DRV_GPIO0_DRV_MASK;
+		if(config.gpio_en)
+			reg_data |= ADPD188_GPIO_DRV_GPIO0_ENA_MASK;
+		else
+			reg_data &= ~ADPD188_GPIO_DRV_GPIO0_ENA_MASK;
+	} else if(config.gpio_id == 1) {
+		if(config.gpio_pol)
+			reg_data |= ADPD188_GPIO_DRV_GPIO1_POL_MASK;
+		else
+			reg_data &= ~ADPD188_GPIO_DRV_GPIO1_POL_MASK;
+		if(config.gpio_drv)
+			reg_data |= ADPD188_GPIO_DRV_GPIO1_DRV_MASK;
+		else
+			reg_data &= ~ADPD188_GPIO_DRV_GPIO1_DRV_MASK;
```

Comment:
don't you need to enable gpio1 also ?

Reply:
The option is not present in the datasheet. It seems that GPIO1 is always enabled. I suspect it is because GPIO0 can also be a clock input and the GPIO driver is disconnected at those times.